tiledb/api/src/query/read/aggregate.rs

```diff
@@ -0,0 +1,665 @@
+        attr_name: String,
+    ) -> TileDBResult<AggregateTypedBuilder<Self, u64>> {
+        self.apply_typed_aggregate::<u64>(AggregateType::NullCount(attr_name))
+    }
+
+    fn mean(
+        self,
+        attr_name: String,
+    ) -> TileDBResult<AggregateTypedBuilder<Self, f64>> {
+        self.apply_typed_aggregate::<f64>(AggregateType::Mean(attr_name))
+    }
+
+    fn sum<T: 'static>(
+        self,
+        attr_name: String,
+    ) -> TileDBResult<AggregateTypedBuilder<Self, T>> {
+        self.apply_typed_aggregate::<T>(AggregateType::Sum(attr_name))
+    }
+
+    fn min<T: 'static>(
+        self,
+        attr_name: String,
+    ) -> TileDBResult<AggregateTypedBuilder<Self, T>> {
+        self.apply_typed_aggregate::<T>(AggregateType::Min(attr_name))
+    }
+
+    fn max<T: 'static>(
+        self,
+        attr_name: String,
+    ) -> TileDBResult<AggregateTypedBuilder<Self, T>> {
+        self.apply_typed_aggregate::<T>(AggregateType::Max(attr_name))
+    }
+}
+
+pub trait AggregateEnumBuilderTrait: QueryBuilder {
+    fn apply_enum_aggregate(
+        self,
+        agg_type: AggregateType,
+    ) -> TileDBResult<AggregateEnumBuilder<Self>>
+    where
+        Self: AggregateBuilderTrait,
+    {
+        let schema = self.base().array().schema()?;
+        let aggregate_result_type = aggregate_type(&agg_type, &schema)?;
+        let agg_builder = match aggregate_result_type {
+            Datatype::UInt8 => AggregateEnumBuilder::UInt8(
+                self.apply_typed_aggregate::<u8>(agg_type)?,
+            ),
+            Datatype::UInt16 => AggregateEnumBuilder::UInt16(
+                self.apply_typed_aggregate::<u16>(agg_type)?,
+            ),
+            Datatype::UInt32 => AggregateEnumBuilder::UInt32(
+                self.apply_typed_aggregate::<u32>(agg_type)?,
+            ),
+            Datatype::UInt64 => AggregateEnumBuilder::UInt64(
+                self.apply_typed_aggregate::<u64>(agg_type)?,
+            ),
+            Datatype::Int8 => AggregateEnumBuilder::Int8(
+                self.apply_typed_aggregate::<i8>(agg_type)?,
+            ),
+            Datatype::Int16 => AggregateEnumBuilder::Int16(
+                self.apply_typed_aggregate::<i16>(agg_type)?,
+            ),
+            Datatype::Int32 => AggregateEnumBuilder::Int32(
+                self.apply_typed_aggregate::<i32>(agg_type)?,
+            ),
+            Datatype::Int64 => AggregateEnumBuilder::Int64(
+                self.apply_typed_aggregate::<i64>(agg_type)?,
+            ),
+            Datatype::Float32 => AggregateEnumBuilder::Float32(
+                self.apply_typed_aggregate::<f32>(agg_type)?,
+            ),
+            Datatype::Float64 => AggregateEnumBuilder::Float64(
+                self.apply_typed_aggregate::<f64>(agg_type)?,
+            ),
+            _ => unreachable!(
+                "aggregate_type function should return a numeric type."
+            ),
+        };
+        Ok(agg_builder)
+    }
+}
+
```

Comment:
I definitely want to see a `mod tests` in here somewhere which covers the error conditions, as there are quite a few of them.